src/Http/Http.Extensions/test/HttpResponseJsonExtensionsTests.cs

```diff
@@ -481,6 +482,83 @@ public async Task WriteAsJsonAsync_NullValue_WithJsonTypeInfo_JsonResponse()
         Assert.Equal("null", data);
     }
 
+    [Fact]
+    public async Task WriteAsJsonAsyncGeneric_AsyncEnumerableStartAsyncNotCalled()
+    {
+        // Arrange
+        var body = new MemoryStream();
+        var context = new DefaultHttpContext();
+        context.Response.Body = body;
+        var responseBodyFeature = new TestHttpResponseBodyFeature(context.Features.GetRequiredFeature<IHttpResponseBodyFeature>());
+        context.Features.Set<IHttpResponseBodyFeature>(responseBodyFeature);
+
+        // Act
+        await context.Response.WriteAsJsonAsync(AsyncEnumerable());
+
+        // Assert
+        Assert.Equal(ContentTypeConstants.JsonContentTypeWithCharset, context.Response.ContentType);
+        Assert.Equal(StatusCodes.Status200OK, context.Response.StatusCode);
+
+        Assert.Equal("[1,2]", Encoding.UTF8.GetString(body.ToArray()));
+
+        async IAsyncEnumerable<int> AsyncEnumerable()
+        {
+            Assert.False(responseBodyFeature.StartCalled);
+            await Task.Yield();
```

Comment:
Wouldn't it be better to test this after yielding?

```suggestion
            await Task.Yield();
            Assert.False(responseBodyFeature.StartCalled);
```

I wonder if we should add an end-to-end test where we try adding a header after yielding.